content/things/rust-to-webassembly.md

```diff
@@ -0,0 +1,637 @@
+
+### Exporting
+
+Okay, we have created a library crate with the default `add` function inside, we have set the crate type to `cdylib` and we have compiled that crate to WebAssembly. Let’s take a look at the WebAssembly code that the compiler generates. For that purpose, I recommend the [WebAssembly Binary Toolkit (wabt for short)][wabt], which provides helpful tools like `wasm-objdump` and `wasm2wat`. It’s also good to have [binaryen] installed which provides `wasm-dis`, but does not emitted "proper" WebAssembly Text Format (WAT). In fact, I usually only use `wasm-opt` from binaryen.
+
+```
+$ wasm2wat ./target/wasm32-unknown-unknown/release/my_project.wasm
+```
+
+This command will convert a WebAssembly binary to WAT and output it to stdout.
+
+```wasm
+(module
+  (table (;0;) 1 1 funcref)
+  (memory (;0;) 16)
+  (global $__stack_pointer (mut i32) (i32.const 1048576))
+  (global (;1;) i32 (i32.const 1048576))
+  (global (;2;) i32 (i32.const 1048576))
+  (export "memory" (memory 0))
+  (export "__data_end" (global 1))
+  (export "__heap_base" (global 2)))
+```
+
+It is with outrage that we discover that our `add` function has been completely removed from the binary. All we are left with is a stack pointer, and two globals designating where the data section ends and the heap starts. Turns out declaring a function as `pub` is _not_ enough to get it to show up in our final WebAssembly module. I kinda wish it was enough, but I suspect `pub` is exclusive about Rust module visibility, not about linker-level visibility.
+
+The quickest way to change the compiler's behavior is to explicitly give the function an export name:
+
+|||codediff|rust
++ #[export_name = "add"]
+  pub fn add(left: usize, right: usize) -> usize {
+      left + right
+  }
+|||
+
+If you don’t intend to change the functions external name, you can also use `#[no_mangle]` instead, which will have the same effect. I feel like `no_mangle` hides the true intention of the developer here, so I tend to prefer the `export_name`. 
+
+Having given our `add` function an export name, we can compile the project again and inspect the resulting WebAssembly file:
+
+|||codediff|wasm
+  (module
++   (type (;0;) (func (param i32 i32) (result i32)))
++   (func $add (type 0) (param i32 i32) (result i32)
++     local.get 1
++     local.get 0
++     i32.add)
+    (table (;0;) 1 1 funcref)
+    (memory (;0;) 16)
+    (global $__stack_pointer (mut i32) (i32.const 1048576))
+    (global (;1;) i32 (i32.const 1048576))
+    (global (;2;) i32 (i32.const 1048576))
+    (export "memory" (memory 0))
++   (export "add" (func $add))
+    (export "__data_end" (global 1))
+    (export "__heap_base" (global 2)))
+|||
+
+This module is easy to run in Node, or Deno or even the browser:
```

Comment:
The usual concern about "easy" in an explanatory blog post. Someone reading it might have never seen how Wasm is instantiated under the hood if they used it via Emscripten or wasm-bindgen, so I feel it's worth spelling out what we do here + MDN links.